homeassistant/components/sensor/utility_meter.py

```diff
@@ -0,0 +1,264 @@
+
+        async_add_entities([meter])
+
+        @callback
+        def async_start_pause_meter(service):
+            """Process service start_pause meter."""
+            for entity in service.data.get(ATTR_ENTITY_ID):
+                if entity in hass.data[DATA_KEY]:
+                    hass.async_add_job(
+                        hass.data[DATA_KEY][entity].async_start_pause_meter())
+
+        @callback
+        def async_reset_meter(service):
+            """Process service reset meter."""
+            for entity in service.data.get(ATTR_ENTITY_ID):
+                if entity in hass.data[DATA_KEY]:
+                    hass.async_add_job(
+                        hass.data[DATA_KEY][entity].async_reset_meter(
+                            dt_util.utcnow()))
+
+        hass.services.async_register(DOMAIN, SERVICE_START_PAUSE,
+                                     async_start_pause_meter,
+                                     schema=SERVICE_METER_SCHEMA)
+        hass.services.async_register(DOMAIN, SERVICE_RESET,
+                                     async_reset_meter,
+                                     schema=SERVICE_METER_SCHEMA)
+
+    # Wait until start event is sent to load this component.
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, run_setup)
+
+
+class UtilityMeterSensor(RestoreEntity):
+    """Representation of an utility meter sensor."""
+
+    def __init__(self, hass, source_entity, name, meter_type, meter_offset=0,
+                 paused=False):
+        """Initialize the min/max sensor."""
+        self._hass = hass
+        self._sensor_source_id = source_entity
+        self._state = 0
+        self._last_period = 0
+        self._last_reset = dt_util.now()
+        # _collecting initializes in inverted logic
+        self._collecting = None if not paused else lambda: None
+        if name:
+            self._name = name
+        else:
+            self._name = '{} meter'.format(source_entity)
+        self._unit_of_measurement = None
+        self._period = meter_type
+        self._period_offset = meter_offset
+
+        if meter_type == HOURLY:
+            async_track_time_change(hass, self.async_reset_meter,
+                                    minute=meter_offset, second=0)
+        if meter_type == DAILY:
+            async_track_time_change(hass, self.async_reset_meter,
+                                    hour=meter_offset, minute=0, second=0)
+        elif meter_type in [WEEKLY, MONTHLY, YEARLY]:
+            async_track_time_change(hass, self._async_reset_meter,
+                                    hour=0, minute=0, second=0)
+
+    @callback
+    def async_reading(self, entity, old_state, new_state):
+        """Handle the sensor state changes."""
+        if old_state is None:
+            return
+
+        if self._unit_of_measurement is None:
+            self._unit_of_measurement = new_state.attributes.get(
+                ATTR_UNIT_OF_MEASUREMENT)
```

Comment:
Should we also check if `new_state` has `ATTR_UNIT_OF_MEASUREMENT`. So we don't go from having a unit of measurement back to not having one.

Reply:
You can't go back to not having a unit of measurement thanks to the `if`
But you could find yourself replacing None with None (fixed)